Add support for newlines, backslashes, trailing comments and unquoted UTF-8

bbc2 opened the pull request:
This adds support for:

* multiline values (i.e. containing newlines or escaped \n), fixes #89
* backslashes in values, fixes #112
* trailing comments, fixes #141
* UTF-8 in unquoted values, fixes #147

This supersedes a previous pull-request, #142, which would add support for
multiline values in `Dotenv.parse` but not in the CLI (`dotenv get` and `dotenv
set`).

The internal change is significant but I have added a lot of test cases to reduce the risk of breaking anything.  Previous test cases are still present, so I wouldn't expect any major backward incompatibility.

I have written detailed commit messages and made my code as clear as possible.  Let me know if anything should be improved.  I'll be happy to fix anything unsatisfactory.

bbc2 commented:
I added a test case so that coverage is kept at a high level.  :chart_with_upwards_trend:

theskumar commented:
This looks great! I'm happy to merge this in, possible you can make it upto date with master? Let me know so I can hold up any further merges to master. 

The conflict seems to be due to https://github.com/theskumar/python-dotenv/commit/43af2c57a3622822fc9f8c09031d2019ef331dd3

bbc2 commented:
I've just rebased, this should be good to go!

I noticed two small issues when rebasing, which I've fixed and added test cases for:
* Parsing could fail when a line would only contain a comment.
* A special character in the name of a variable (it's legitimate but very unusual) would cause the parser to loop forever.

OlegSmelov commented on dotenv/main.py:
```diff
@@ -2,47 +2,91 @@
-        return None, None
+            \s*                     # trailing whitespace
+            (?:\#[^\r\n]*)?         # comment
+            \s*                     # trailing whitespace
```
Isn't this line redundant? If a comment is present, `[^\r\n]*` can capture the whitespace, if not, the previous `\s*` will capture it.

amancevice commented:
Any plans to release soon with this patch included?

theskumar commented:
Thanks @bbc2 and others for the patience. I've had some little busy schedule. I'm updating the README etc. and then make a release.

mungojam commented:
This was a breaking change for us. We had an entry `output = "C:\temp"` in our .env file which was then broken by this change as the `\t` is now interpreted as a tab character. We didn't pick up on it initially because VS Code doesn't apply this new behaviour.

mungojam commented:
I've avoided the issue by removing the speech marks now. That seems to deal fine with spaces so I shouldn't really need to use speech marks for file paths

ChrisC413 commented:
This was a breaking change for me. I use a local .env file to store a password for local testing, the password contains a #

bbc2 commented:
I'm sorry it broke your use cases.  I'll try to fix this soon. My plan:
* Add test cases to ensure we don't break things again by accident.
* Update the changelog to notify users that the new version can break some .env files.
* Open an issue to determine how python-dotenv should parse and interpret those cases in the future.
